### DISCUSSION TITLE
[Guidance] Data type conversion in dialect?

### DISCUSSION THREAD

**Discussion by fdcastel** [sqlalchemy/sqlalchemy#9661] [Usage Questions]
I'm working on a revamp for Firebird SQLAlchemy 2.0 dialect but I'm banging my head with this test:

```python
class DateTest(_DateFixture, fixtures.TablesTest):
    __requires__ = ("date",)
    __backend__ = True
    datatype = Date
    data = datetime.date(2012, 10, 15)

    @testing.requires.date_implicit_bound
    def test_select_direct(self, connection):
        result = connection.scalar(select(literal(self.data)))
        eq_(result, self.data)
```

Whatever I do I'm always getting `result` as a string instead of a date:

```
> result
'2012-10-15'

> self.data
datetime.date(2012, 10, 15)
```

I've made a `_FBDate` type as:

```python
class _FBDate(DATE):
    def bind_processor(self, dialect):
        def process(value):
            return datetime.date(value.year, value.month, value.day) if isinstance(value, datetime.date) else value
        return process

    def result_processor(self, dialect, coltype):
        def str_to_date(value: Optional[str]) -> Optional[datetime.date]:
            return datetime.date.fromisoformat(value) if value is not None else None 
        return str_to_date
```

I've registered it in `colspecs` and `ischema_names`:

```python
colspecs = {
    sqltypes.DATE: _FBDate
}

ischema_names = {
    # ...
    "DATE": _FBDate,
    # ...
}
```

I'm surely missing something out, but I can't find what it is just studying the code.

Using `fdb` firebird driver. Later I intend to create another dialect using `firebird-driver` (a new, more modern driver).

**Comment by zzzeek** [comment]
the "date_implicit_bound" req refers to a driver level behavior that is often not present, which is that if you run a plain statement like "SELECT :param", pass in a Python date() object, and get the result, the cursor.description will know that the value is a date.  you would not want to have to use a result_processor() for this test if the DBAPI actually supports Python date() values directly.   so if fdb isn't doing that here, you can skip it.  None of the MySQL drivers support this test either.

all of that said, if result_processor() is not being called at all you'd want to ensure that `_FBDate` is configured in the dialect's `colspecs` dictionary.   The SQLite dialect is an example of a dialect that receives dates back as strings and runs result_processor() in all cases if you want to look at that.

**Reply by fdcastel** [reply]
Thanks @zzzeek. To clarify: the dialect `colspecs` property **IS** set (with the value of global `colspecs`):

```python
colspecs = {
    sqltypes.DATE: _FBDate
}

class FBDialect(default.DefaultDialect):
    name = "firebird"
    # ...
    colspecs = colspecs
    ischema_names = ischema_names
    # ...
```

but neither `bind_processor()` nor `result_processor()` from `_FBDate` are being called now (I could swear they were, yesterday!).

Any other pointers for me to check?

**Reply by zzzeek** [reply]
I would pdb into default.py in [cached_result_processor](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/engine/default.py#L1545), wait for "colname" to be something that should have a type, and see what's being returned.

**Comment by fdcastel** [comment]
About `date_implicit_bound`: thanks for the explanation.

Now [this code](https://github.com/sqlalchemy/sqlalchemy/blob/acf7fbd60b9b1291dfc91438416867c88e94c5ba/test/requirements.py#L1204) is clearer to me.
